logging: panic not printed to log file

I'm running with `--logtostderr`. A panic seems to make it to the console but not to the log file.

The log file ends like this:
```
[andrei:~/work/src/github.com/cockroachdb/cockroach] e915821(+13/-1)+ 4s ± tail cockroach-data/logs/cockroach.log

E170929 03:09:41.434035 543 sql/session.go:1370  [client=[::1]:59084,user=root,n1] a SQL panic has occurred!
E170929 03:09:41.434211 543 util/log/crash_reporting.go:117  [n1] a panic has occurred!
```

The console shows:
```
E170929 03:09:41.434035 543 sql/session.go:1370  [client=[::1]:59084,user=root,n1] a SQL panic has occurred!
E170929 03:09:41.434211 543 util/log/crash_reporting.go:117  [n1] a panic has occurred!
panic: unexpected txnState when cleaning up: NoTxn [recovered]
	panic: panic while executing "ROLLBACK": unexpected txnState when cleaning up: NoTxn [recovered]
	panic: panic while executing "ROLLBACK": unexpected txnState when cleaning up: NoTxn [recovered]
	panic: panic while executing "ROLLBACK": unexpected txnState when cleaning up: NoTxn

goroutine 543 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4207b0000, 0x63508e0, 0xc42052cc60)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:183 +0x131
panic(0x5596540, 0xc422152400)
"cockroach-data/logs/cockroach.log" 27693L, 10387088C
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc421520c00, 0x63508e0, 0xc421453d10)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:423 +0xa1
panic(0x5596540, 0xc422152400)
	/Users/andrei/work/src/go/src/runtime/panic.go:491 +0x283
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc42156b000, 0x575e951, 0x9, 0xc42109b67c, 0x8)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1382 +0x262
panic(0x54dad00, 0xc420b537c0)
	/Users/andrei/work/src/go/src/runtime/panic.go:491 +0x283
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn.func1(0xc42303aae8, 0xc42156b090, 0xc420012fc0, 0x0, 0xc4241fe500, 0xc42156b000)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1559 +0x558
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc420012fc0, 0xc42156b000, 0x63530a0, 0x686e348, 0x0, 0x0, 0x0, 0x14e8b65bf58a4bea, 0x1, 0xc4241c9680, ...)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1649 +0xd58
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execSingleStatement(0xc420012fc0, 0xc42156b000, 0x63530a0, 0x686e348, 0x0, 0x0, 0x0, 0x14e8b65bf58a4bea, 0x1, 0xc4241c9680, ...)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1320 +0x6a5
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc420012fc0, 0xc42156b000, 0xc4241fe440, 0x1, 0x1, 0x0, 0x2, 0x6350000, 0x0, 0x6355b00, ...)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1195 +0x3e5
github.com/cockroachdb/cockroach/pkg/sql.runWithAutoRetry(0xc420012fc0, 0xc42156b000, 0xc4241fe440, 0x1, 0x1, 0x0, 0x0, 0x0, 0xfffffffe11293d00, 0xc4239e2b48, ...)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1021 +0x350
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed(0xc420012fc0, 0xc42156b000, 0xc4241fe440, 0x1, 0x1, 0x0, 0x0, 0xc424a10000, 0x9)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:840 +0x130
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc420012fc0, 0xc42156b000, 0xc42109b67c, 0x8, 0x0, 0x0, 0x8, 0xc42109b67c)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:752 +0x49c
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc420012fc0, 0xc42156b000, 0xc42109b67c, 0x8, 0x0, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:638 +0x1bd
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc421520c00, 0xc421520c28, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:612 +0x1d0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc421520c00, 0x63508e0, 0xc421453d10, 0xc4202bf780, 0x5400, 0xc42072c368, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:503 +0x8ae
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42072c200, 0x6350820, 0xc421540300, 0x635ca00, 0xc4207d5a40, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x929
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x635ca00, 0xc4207d5a40)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/server/server.go:748 +0x149
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4207b0000, 0x63508e0, 0xc42052cc60, 0xc42000e048, 0x635ca00, 0xc4207d5a40, 0xc420f9a060)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:140 +0xa9
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:137 +0x263
```

--- a/pkg/cli/interactive_tests/test_missing_log_output.tcl
+++ b/pkg/cli/interactive_tests/test_missing_log_output.tcl
@@ -70,6 +70,28 @@ end_test
 interrupt
 eexpect ":/# "
 
+start_test "Check that panic reports are printed to the log even when --logtostderr is specified"
+send "$argv start -s=path=logs/db --insecure --logtostderr\r"
+eexpect "CockroachDB node starting"
+
+system "$argv sql --insecure -e \"select crdb_internal.force_panic('helloworld')\" || true"
+# Check the panic is reported on the server's stderr
+eexpect "panic: helloworld"
+eexpect "panic while executing"
+eexpect "goroutine"
+eexpect ":/# "
+# Check the panic is reported on the server log file
+send "cat logs/db/logs/cockroach.log\r"
+eexpect "a panic has occurred"
+eexpect "panic while executing"
+eexpect "helloworld"
+eexpect "goroutine"
+eexpect ":/# "
+
+end_test
+
+
+
 start_server $argv
 
 start_test "Test that quit does not emit unwanted logging output"

--- a/pkg/util/log/clog.go
+++ b/pkg/util/log/clog.go
@@ -29,6 +29,7 @@ import (
 	"path/filepath"
 	"regexp"
 	"runtime"
+	"runtime/debug"
 	"strconv"
 	"strings"
 	"sync/atomic"
@@ -710,6 +711,26 @@ func (l *loggingT) putBuffer(b *buffer) {
 	l.freeListMu.Unlock()
 }
 
+// ensureFile ensures that l.file is set and valid.
+func (l *loggingT) ensureFile() error {
+	if l.file == nil {
+		return l.createFile()
+	}
+	return nil
+}
+
+// writeToFile writes to the file and applies the synchronization policy.
+func (l *loggingT) writeToFile(data []byte) error {
+	if _, err := l.file.Write(data); err != nil {
+		return err
+	}
+	if l.syncWrites {
+		_ = l.file.Flush()
+		_ = l.file.Sync()
+	}
+	return nil
+}
+
 // outputLogEntry marshals a log entry proto into bytes, and writes
 // the data to the log files. If a trace location is set, stack traces
 // are added to the entry before marshaling.
@@ -726,7 +747,6 @@ func (l *loggingT) outputLogEntry(s Severity, file string, line int, msg string)
 	// TODO(tschottdorf): this is a pretty horrible critical section.
 	l.mu.Lock()
 
-	// On fatal log, set all stacks.
 	var stacks []byte
 	if s == Severity_FATAL {
 		switch traceback {
@@ -748,38 +768,57 @@ func (l *loggingT) outputLogEntry(s Severity, file string, line int, msg string)
 		l.outputToStderr(entry, stacks)
 	}
 	if logDir.isSet() && s >= l.fileThreshold.get() {
-		if l.file == nil {
-			if err := l.createFile(); err != nil {
-				// Make sure the message appears somewhere.
-				l.outputToStderr(entry, stacks)
-				l.exitLocked(err)
-				l.mu.Unlock()
-				return
-			}
+		if err := l.ensureFile(); err != nil {
+			// Make sure the message appears somewhere.
+			l.outputToStderr(entry, stacks)
+			l.exitLocked(err)
+			l.mu.Unlock()
+			return
 		}
 
 		buf := l.processForFile(entry, stacks)
 		data := buf.Bytes()
 
-		if _, err := l.file.Write(data); err != nil {
+		if err := l.writeToFile(data); err != nil {
 			l.exitLocked(err)
 			l.mu.Unlock()
 			return
 		}
-		if l.syncWrites {
-			_ = l.file.Flush()
-			_ = l.file.Sync()
-		}
 
 		l.putBuffer(buf)
 	}
-	exitFunc := l.exitFunc
+	exitFunc := l.exitFunc // restore the exitFunc
 	l.mu.Unlock()
 	// Flush and exit on fatal logging.
 	if s == Severity_FATAL {
 		// If we got here via Exit rather than Fatal, print no stacks.
 		timeoutFlush(10 * time.Second)
 		exitFunc(255) // C++ uses -1, which is silly because it's anded with 255 anyway.
+	}
+}
+
+// printPanicToFile copies the panic details to the log file. This is
+// useful when the standard error is not redirected to the log file
+// (!stderrRedirected), as the go runtime will only print panics to
+// stderr.
+func (l *loggingT) printPanicToFile(r interface{}) {
+	if !logDir.isSet() {
+		// There's no log file. Nothing to do.
+		return
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if err := l.ensureFile(); err != nil {
+		fmt.Fprintf(OrigStderr, "log: %v", err)
+		return
+	}
+
+	panicBytes := []byte(fmt.Sprintf("%v\n\n%s\n", r, debug.Stack()))
+	if err := l.writeToFile(panicBytes); err != nil {
+		fmt.Fprintf(OrigStderr, "log: %v", err)
+		return
 	}
 }
 

--- a/pkg/util/log/crash_reporting.go
+++ b/pkg/util/log/crash_reporting.go
@@ -118,22 +118,27 @@ func Safe(v interface{}) SafeType {
 func ReportPanic(ctx context.Context, sv *settings.Values, r interface{}, depth int) {
 	Shout(ctx, Severity_ERROR, "a panic has occurred!")
 
+	if stderrRedirected {
+		// We do not use Shout() to print the panic details here, because
+		// if stderr is not redirected (e.g. when logging to file is
+		// disabled) Shout() would copy its argument to stderr
+		// unconditionally, and we don't want that: Go's runtime system
+		// already unconditonally copies the panic details to stderr.
+		// Instead, we copy manually the details to stderr, only when stderr
+		// is redirected to a file otherwise.
+		fmt.Fprintf(OrigStderr, "%v\n\n%s\n", r, debug.Stack())
+	} else {
+		// If stderr is not redirected, then Go's runtime will only print
+		// out the panic details to the original stderr, and we'll miss a copy
+		// in the log file. Produce it here.
+		logging.printPanicToFile(r)
+	}
+
 	SendCrashReport(ctx, sv, depth+1, "", []interface{}{r})
 
 	// Ensure that the logs are flushed before letting a panic
 	// terminate the server.
 	Flush()
-
-	// We do not use Shout() to print the panic details here, because
-	// if stderr is not redirected (e.g. when logging to file is
-	// disabled) Shout() would copy its argument to stderr
-	// unconditionally, and we don't want that: Go's runtime system
-	// already unconditonally copies the panic details to stderr.
-	// Instead, we copy manually the details to stderr, only when stderr
-	// is redirected to a file otherwise.
-	if stderrRedirected {
-		fmt.Fprintf(OrigStderr, "%v\n\n%s\n", r, debug.Stack())
-	}
 }
 
 var crashReportURL = func() string {

